prov/rxm/src/rxm_cq.c

```diff
@@ -708,6 +708,75 @@ static inline void rxm_cq_repost_rx_buffers(struct rxm_ep *rxm_ep)
 	}
 }
 
+static int rxm_cq_reprocess_directed_recvs(struct rxm_recv_queue *recv_queue)
+{
+	struct rxm_rx_buf *rx_buf;
+	struct dlist_entry *entry, *tmp_entry;
+	struct rxm_recv_match_attr match_attr;
+	int ret = 0, count = 0;
+
+	recv_queue->rxm_ep->res_fastlock_acquire(&recv_queue->lock);
+
+	dlist_foreach_container_safe(&recv_queue->unexp_msg_list,
+				     struct rxm_rx_buf, rx_buf,
+				     unexp_msg.entry, tmp_entry) {
+		if (rx_buf->unexp_msg.addr == rx_buf->conn->handle.fi_addr)
+			continue;
+
+		assert(rx_buf->unexp_msg.addr == FI_ADDR_NOTAVAIL);
+
+		match_attr.addr = rx_buf->unexp_msg.addr =
+			rx_buf->conn->handle.fi_addr;
+		match_attr.tag = rx_buf->unexp_msg.tag;
+
+		entry = dlist_remove_first_match(&recv_queue->recv_list,
+						 recv_queue->match_recv,
+						 &match_attr);
+		if (!entry)
+			continue;
+
+		dlist_remove(&rx_buf->unexp_msg.entry);
+		rx_buf->recv_entry = container_of(entry, struct rxm_recv_entry,
+						  entry);
+		recv_queue->rxm_ep->res_fastlock_release(&recv_queue->lock);
```

Comment:
We can't drop the lock on the list without risk of corrupting it.  dlist_foreach_safe() allows removing an entry within the loop, but is not thread safe.  Modifications to the list by another thread can corrupt the list.  It's best to move rx_buf to a local list that we iterate over after walking the unexpected list.

Reply:
Good catch! Done.